kong/db/declarative/init.lua

```diff
@@ -837,175 +834,117 @@ function declarative.load_into_cache(entities, meta, hash, shadow)
-      return nil, err
-    end
-
-    local worker_events = kong.worker_events
+    elseif err:find("MDB_MAP_FULL", nil, true) then
+      return nil, "map full"
 
-    -- ensure any previous update finished (we're flipped to the latest page)
-    ok, err = worker_events.poll()
-    if not ok then
-      kong_shm:delete(DECLARATIVE_LOCK_KEY)
+    else
       return nil, err
     end
 
-    if SUBSYS == "http" and #kong.configuration.stream_listeners > 0 and
-       get_phase() ~= "init_worker"
-    then
+    if SUBSYS == "http" and #kong.configuration.stream_listeners > 0 then
       -- update stream if necessary
-      -- TODO: remove this once shdict can be shared between subsystems
 
       local sock = ngx_socket_tcp()
       ok, err = sock:connect("unix:" .. PREFIX .. "/stream_config.sock")
       if not ok then
-        kong_shm:delete(DECLARATIVE_LOCK_KEY)
         return nil, err
       end
 
-      local json = cjson.encode({ entities, meta, hash, })
       local bytes
-      bytes, err = sock:send(json)
+      bytes, err = sock:send(default_ws)
       sock:close()
 
       if not bytes then
-        kong_shm:delete(DECLARATIVE_LOCK_KEY)
         return nil, err
       end
 
-      assert(bytes == #json, "incomplete config sent to the stream subsystem")
-    end
-
-    if exiting() then
-      kong_shm:delete(DECLARATIVE_LOCK_KEY)
-      return nil, "exiting"
-    end
-
-    local default_ws
-    ok, err, default_ws = declarative.load_into_cache(entities, meta, hash, SHADOW)
-    if ok then
-      ok, err = worker_events.post("declarative", "flip_config", default_ws)
-      if ok ~= "done" then
-        kong_shm:delete(DECLARATIVE_LOCK_KEY)
-        return nil, "failed to flip declarative config cache pages: " .. (err or ok)
-      end
-
-    else
-      kong_shm:delete(DECLARATIVE_LOCK_KEY)
-      return nil, err
+      assert(bytes == #default_ws,
+             "incomplete default workspace id sent to the stream subsystem")
     end
 
-    ok, err = kong_shm:set(DECLARATIVE_PAGE_KEY, kong.cache:get_page())
-    if not ok then
-      kong_shm:delete(DECLARATIVE_LOCK_KEY)
-      return nil, "failed to persist cache page number: " .. err
-    end
 
     if exiting() then
-      kong_shm:delete(DECLARATIVE_LOCK_KEY)
       return nil, "exiting"
     end
 
-    local sleep_left = DECLARATIVE_LOCK_TTL
-    local sleep_time = 0.0375
-
-    while sleep_left > 0 do
-      local flips = kong_shm:get(DECLARATIVE_LOCK_KEY)
-      if flips == nil or flips >= WORKER_COUNT then
-        break
-      end
+    return true
+  end
 
-      sleep_time = sleep_time * 2
-      if sleep_time > sleep_left then
-        sleep_time = sleep_left
-      end
+  -- If it takes more than 60s it is very likely to be an internal error.
+  -- However it will be reported as: "failed to broadcast reconfigure event: recursive".
+  -- Let's paste the error message here in case someday we try to search it.
+  -- Should we handle this case specially?
```

Comment:
As somebody who has seen this specific error message on a production system, big big +1 to anything that we can do to make the error more meaningful and explanatory.

Reply:
I am not a 100% sure about the triggering condition of this, and I hesitate to make a change without understanding why the "recursive" error occurs. Will try to syncup with @Suika-Kong more offline on this error handling after it is merged (the current PR is getting hard to manage).